File: Mpi/Mpi.cs
using System;
using System.Numerics;
using System.Text;

using static Extreme.Parallel.UnsafeNativeMethods;
using UNM = Extreme.Parallel.UnsafeNativeMethods;
using System.ComponentModel.Design;
using System.Runtime.Remoting.Messaging;
using System.Runtime.ConstrainedExecution;

namespace Extreme.Parallel
{
    public unsafe class Mpi : IDisposable
    {
        public readonly int Master = 0;

        private static readonly IntPtr CommWorld = GetCommWorld();

		private static readonly IntPtr CommNull =GetCommNull();

        public static readonly IntPtr Int = GetMpiInt();
        public static readonly IntPtr Float = GetMpiFloat();
        public static readonly IntPtr Double = GetMpiDouble();
        public static readonly IntPtr Complex = GetMpiDoubleComplex();
        public static readonly int AnySource = GetMpiAnySource();
		public static readonly IntPtr InPlace=GetMpiInPlace();
        public static readonly IntPtr MpiOpSumm = GetMpiOpSum();


        private readonly int _rank = 0;
        private readonly int _size = 1;
        private Mpi()
        {
            UnsafeNativeMethods.Init();


			IntPtr comm;
			Communicator = CommWorld;

			int rank;
			int size;

			UnsafeNativeMethods.GetCommRank (Communicator,&rank);
			UnsafeNativeMethods.GetCommSize (Communicator,&size);
			_rank = rank;
			_size = size;
		}


        private static bool _initialized = false;

        public static Mpi Init()
        {
            if (_initialized)
                throw new InvalidOperationException("Mpi service is already initialized");

            try
            {
                return new Mpi();
            }
            finally
            {
                _initialized = true;
            }
        }
		private Mpi(IntPtr comm)
		{
			if (_initialized) {
				Communicator = comm;

				int rank;
				int size;
				int err;
				err = UnsafeNativeMethods.GetCommRank (comm, &rank);
					
				if (err != 0)
					throw new InvalidOperationException (GetErrorString (err));
				_rank = rank;

				err = UnsafeNativeMethods.GetCommSize (comm, &size);
				if (err != 0)
					throw new InvalidOperationException (GetErrorString (err));

				_size = size;
			} else {
				
			}
		}
		public Mpi Merge (params int[] ranks)
		{
			var comm=CreateNewCommunicator(ranks);
			if (comm != CommNull)
				return new Mpi (comm);
			else
				return null;
		}

		public Mpi Separate(int color){
			IntPtr comm;
			UNM.CommSplit (Communicator, color, Rank, &comm);
			return new Mpi (comm);
		}


        public bool IsParallel => Size > 1;
        public int Size => _size;
        public int Rank => _rank;
        public bool IsMaster => _rank == Master;
		public IntPtr Communicator { get; private set; }=CommNull;

        private static void WithErrorHandling(int err)
        {
            if (err != 0)
                throw new InvalidOperationException(GetErrorString(err));
        }

        private IntPtr CreateNewCommunicator(params int[] ranks)
        {
            IntPtr group;
            CommGroup(Communicator, &@group);

            IntPtr newGroup;
            GroupIncl(@group, ranks.Length, ranks, &newGroup);

            IntPtr newComm;
			CommCreate(Communicator, newGroup, &newComm);

            return newComm;
        }


		public Mpi Dup(){
			IntPtr comm;

			UNM.CommDup (this.Communicator, &comm);
			return new Mpi (comm);
		}

        public Complex AllReduce(Complex value)
        {
            Complex result;
			UnsafeNativeMethods.AllReduce(&value, &result, 1, Complex, Communicator);

            return result;
        }

        public double AllReduce(double value)
        {
            double result;
			UnsafeNativeMethods.AllReduce(&value, &result, 1, Double, Communicator);
            return result;
        }

		public int AllReduce(int value)
		{
			int result;
			UnsafeNativeMethods.AllReduce(&value, &result, 1, Int, Communicator);
			return result;
		}


        public void Reduce(double* value, double* result, int length)
        {
			if (result == value && IsMaster) {
				UnsafeNativeMethods.Reduce ((void*)InPlace, result, length, Double, Communicator);
			} else {
				UnsafeNativeMethods.Reduce (value, result, length, Double, Communicator);
			}
        }

		public void LogicalReduce(double* value, double* result, int length)
		{

			if (result == value && IsMaster) {
				UnsafeNativeMethods.LogicalReduce ((void*)InPlace, result, length,  Communicator);
			} else {
				UnsafeNativeMethods.LogicalReduce (value, result, length,  Communicator);
			}
		}

		public static long CommunicatorC2Fortran(IntPtr comm)
		{
			return UnsafeNativeMethods.CommunicatorC2Fortran(comm);
		}

		public long CommunicatorC2Fortran()
		{
			return CommunicatorC2Fortran(this.Communicator);
		}

		public static void Barrier(IntPtr comm)
            => WithErrorHandling(UnsafeNativeMethods.Barrier(comm));

		public void Barrier()
		=> WithErrorHandling(UnsafeNativeMethods.Barrier(Communicator));

        public int BroadCast(int root, int value)
        {
			Bcast(&value, 1, Int, root, Communicator);
           return value;
        }

        public void AllGatherV(Complex* src, int sendSize, Complex* dst, int[] rCounts, int[] rDispl)
        {
            fixed (int* cntPtr = &rCounts[0], dsplPtr = &rDispl[0])
			UnsafeNativeMethods.AllGatherV(src, sendSize, dst, cntPtr, dsplPtr,Communicator);
        }

        public void GatherV(Complex* src, int sendSize, Complex* dst, int[] rCounts, int[] rDispl)
        {
            fixed (int* cntPtr = &rCounts[0], dsplPtr = &rDispl[0])
			UnsafeNativeMethods.GatherV(src, sendSize, dst, cntPtr, dsplPtr,Communicator);
        }

        public void BroadCast( int root, double[] values)
        {
            fixed (double* ptr = &values[0])
			Bcast(ptr, values.Length, Double, root, Communicator);
        }

		public void BroadCast( int root, int[,,] values)
		{
			fixed (int* ptr = &values[0,0,0])
				BroadCast ( root, ptr, values.Length);
		}


        public void BroadCast(int root, Complex[] values)
        {
            fixed (Complex* ptr = &values[0])
			Bcast(ptr, values.Length, Complex, root, Communicator);
        }

        public void BroadCast(int root, double* values, int length)
        {
            Bcast(values, length, Double, root, Communicator);
        }


        public void BroadCast(int root, Complex* values, int length)
        {
			Bcast(values, length, Complex, root, Communicator);
        }

		public void BroadCast( int root, int* values, int length)
		{
			Bcast(values, length, Int, root, Communicator);
		}



        public int Send(void* data, int count, IntPtr datatype, int dest, int tag)
		=> UNM.Send(data, count, datatype, dest, tag, Communicator);

        public int Recv(void* data, int count, IntPtr datatype, int source, int tag, out int actualSource)
		=> UNM.Recv(data, count, datatype, source, tag, Communicator, out actualSource);

        public static string GetErrorString(int error)
            => UNM.GetErrorString(error);

        public string GetProcessorName()
        {
            var name = new byte[GetMaxProcessorName()];
            int length = 0;

            fixed (byte* namePtr = &name[0])
            {
                int err = UNM.GetProcessorName(namePtr, ref length);
            }

            return Encoding.ASCII.GetString(name, 0, length);
        }

	
        public void Gather(Complex* dst, Complex* src, int dstSize, int srcSize)
        {
			UNM.Gather(src, srcSize, dst, dstSize, Master, Communicator);
        }

        public void Gather(Complex[] dst, Complex src)
        {
            fixed (Complex* dstPtr = &dst[0])
				UNM.Gather(&src, 1, dstPtr, 1, Master, Communicator);
        }

        public void AllToAll(Complex* buffer, int size)
		=> AllToAllDoubleComplexInPlace(buffer, size, Communicator);

        public void AllToAll(Complex* src, int srcSize, Complex* dst, int dstSize)
		=> AllToAllDoubleComplex(src, srcSize, dst, dstSize, Communicator);


		public void Free(){
			IntPtr comm = Communicator;
			if (Communicator!=CommNull&&Communicator!=CommWorld)
				UnsafeNativeMethods.CommFree(&comm);
			Communicator = CommNull;
		}

        public void Dispose()
        {
           // FinalizeMpi();
			if (Communicator == CommWorld)
				Finalize ();
			Free ();
        }

		private static void Finalize()
		{
			FinalizeMpi ();
		}
    }
}

File: Native/Native.MpiWrapper/mpi.c
//Copyright (c) 2016 by ETH Zurich, Alexey Geraskin, Mikhail Kruglyakov, and Alexey Kuvshinov
#include <stddef.h>
#include <mpi.h>

#ifdef WINDOWS
#define DLLEXPORT __declspec(dllexport)
#else
#define DLLEXPORT 
#endif

#ifdef MSMPI
#define MpiType long long 
#endif

#ifdef OpenMPI
#define MpiType void*
#endif

#ifdef MPICH
#define MpiType long long
#endif

typedef struct { double re; double im; } complex16;

DLLEXPORT MPI_Op GetMpiOpSum() { return MPI_SUM; }

DLLEXPORT MpiType GetCommWorld() { return MPI_COMM_WORLD; }

DLLEXPORT MpiType GetCommNull() { return MPI_COMM_NULL; }

DLLEXPORT MpiType GetMpiInt() { return MPI_INT; }

DLLEXPORT MpiType GetMpiFloat() { return MPI_FLOAT; }
DLLEXPORT MpiType GetMpiDouble() { return MPI_DOUBLE; }
DLLEXPORT MpiType GetMpiDoubleComplex() { return MPI_C_DOUBLE_COMPLEX; }


DLLEXPORT int GetMaxProcessorName() { return MPI_MAX_PROCESSOR_NAME; }
DLLEXPORT MpiType GetMpiAnySource() { return MPI_ANY_SOURCE; }
DLLEXPORT void* GetMpiInPlace(){ return MPI_IN_PLACE; }

DLLEXPORT int GetErrorString(int errorcode, char *string, int *resultlen)
{
	return MPI_Error_string(errorcode, string, resultlen);
}
DLLEXPORT void GetThreadSupportLevels(int* sup_level) { 
	sup_level[0]=MPI_THREAD_SINGLE;
	sup_level[1]=MPI_THREAD_FUNNELED;
	sup_level[2]=MPI_THREAD_SERIALIZED;
	sup_level[3]=MPI_THREAD_MULTIPLE;
  }

DLLEXPORT int Init(int* thread_support)
{
	int provided;
	int err;
	err=MPI_Init_thread(NULL, NULL,   MPI_THREAD_MULTIPLE, thread_support);

	return err;
}

DLLEXPORT long long CommunicatorC2Fortran(MPI_Comm comm)
{
	return  MPI_Comm_c2f(comm);
}

DLLEXPORT int CommSplit(MPI_Comm comm, int color, int key, MPI_Comm *newcomm){
	return MPI_Comm_split(comm, color,  key,	newcomm);
    }

DLLEXPORT int CommDup(MPI_Comm comm, MPI_Comm *newcomm){
	return MPI_Comm_dup(comm,  newcomm);
}

DLLEXPORT int Barrier(MPI_Comm comm)
{
	return MPI_Barrier(comm);
}

DLLEXPORT int AllGatherV(void* sendbuf, int size, void* rbuf, int *recvcounts, int *displs, MPI_Comm comm)
{
	return MPI_Allgatherv(sendbuf, size, MPI_DOUBLE_COMPLEX, rbuf, recvcounts, displs, MPI_DOUBLE_COMPLEX, comm);
}

DLLEXPORT int Gather(void *sendbuf, int sendcount, void *recvbuf, int recvcount, int root, MPI_Comm comm)
{
	return MPI_Gather(sendbuf, sendcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, root, comm);
}

DLLEXPORT int GatherV(void* sendbuf, int size, void* rbuf, int *recvcounts, int *displs, MPI_Comm comm)
{
	return MPI_Gatherv(sendbuf, size, MPI_DOUBLE_COMPLEX, rbuf, recvcounts, displs, MPI_DOUBLE_COMPLEX, 0, comm);
}

DLLEXPORT int AllReduce(void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Comm comm)
{
	return MPI_Allreduce(sendbuf, recvbuf, count, datatype, MPI_SUM, comm);
}

DLLEXPORT int Reduce(void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Comm comm)
{
	return MPI_Reduce(sendbuf, recvbuf, count, datatype, MPI_SUM, 0, comm);
}

//-----------------------------------
void PositiveDoubleLogicalOr( double *in, double *inout, int *len, MPI_Datatype *dptr )
    {
        int i;
        double c;

    	for (i=0; i< *len; i++) {
    		c=in[i] < 0 ? inout[i] : in[i];
            inout[i] = c;
        }
    }

//-------------------------------
DLLEXPORT int LogicalReduce(void *sendbuf, void *recvbuf, int count, MPI_Comm comm)
{
	MPI_Op logical_reduce;
	int error;
	int commutative=1;
	MPI_Op_create( PositiveDoubleLogicalOr, commutative, &logical_reduce );
	
	error=MPI_Reduce(sendbuf, recvbuf, count, MPI_DOUBLE, logical_reduce, 0, comm);
	MPI_Op_free(&logical_reduce);
	return error; 
}


DLLEXPORT int GetProcessorName(char *name, int *resultlen)
{
	return MPI_Get_processor_name(name, resultlen);
}


DLLEXPORT int GetCommWorldRank(int* rank)
{
	return MPI_Comm_rank(MPI_COMM_WORLD, rank);
}

DLLEXPORT int GetCommWorldSize(int* size)
{
	return MPI_Comm_size(MPI_COMM_WORLD, size);
}



DLLEXPORT int GetCommRank(MPI_Comm comm, int* rank)
{
	return MPI_Comm_rank(comm, rank);
}

DLLEXPORT int GetCommSize(MPI_Comm comm,int* size)
{
	return MPI_Comm_size(comm, size);
}



DLLEXPORT int CommCreate(MPI_Comm comm, MPI_Group group, MPI_Comm *newcomm)
{
	return MPI_Comm_create(comm, group, newcomm);
}

DLLEXPORT int GroupIncl(MPI_Group group, int n, int ranks[], MPI_Group *newgroup)
{
	return MPI_Group_incl(group, n, ranks, newgroup);
}

DLLEXPORT int CommGroup(MPI_Comm comm, MPI_Group *group)
{
	return MPI_Comm_group(comm, group);
}

DLLEXPORT int Bcast(void *buffer, int count, MPI_Datatype datatype, int root, MPI_Comm comm)
{
	return	MPI_Bcast(buffer, count, datatype, root, comm);
}


DLLEXPORT int Scatter(void* sendbuf,void* recvbuf, int count, MPI_Datatype datatype, int root, MPI_Comm comm)
{
	return MPI_Scatter(sendbuf, count, datatype, recvbuf, count, datatype,  root, comm);
}

DLLEXPORT int Send(void* data, int count, MPI_Datatype datatype, int dest, int tag, MPI_Comm comm)
{
	return MPI_Send(data, count, datatype, dest, tag, comm);
}

DLLEXPORT int Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag, MPI_Comm comm, int* actualSource)
{
	MPI_Status status;
	int err = MPI_Recv(buf, count, datatype, source, tag, comm, &status);
	(*actualSource) = status.MPI_SOURCE;
	return err;
}

DLLEXPORT int SendComplexMatrix(void* data, int nx, int ny, int destination, int tag, MPI_Comm comm)
{
	return MPI_Send(data, nx*ny, MPI_DOUBLE_COMPLEX, destination, tag, comm);
}

DLLEXPORT int AllToAllDoubleComplex(complex16 *sendbuf, int sendcount, void *recvbuf, int recvcount, MPI_Comm comm)
{
	return MPI_Alltoall(sendbuf, sendcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, comm);
}

DLLEXPORT int AllToAllDoubleComplexInPlace(void *recvbuf, int recvcount, MPI_Comm comm)
{
	return MPI_Alltoall(MPI_IN_PLACE, recvcount, MPI_DOUBLE_COMPLEX, recvbuf, recvcount, MPI_DOUBLE_COMPLEX, comm);
}


DLLEXPORT int SendRecv(void* sendbuf,void* recvbuf ,int count,int source, int dest, int tag,MPI_Datatype datatype,  MPI_Comm comm)
{
	return  MPI_Sendrecv(sendbuf, count, MPI_INT, dest, tag, recvbuf, count, datatype,  source, tag,comm, MPI_STATUS_IGNORE);
}

DLLEXPORT int CommFree(MPI_Comm* comm){
	return  MPI_Comm_free(comm);
}

int MPI_Comm_free(
  MPI_Comm *comm
);


DLLEXPORT void Finalize()
{
	MPI_Finalize();
}

    
